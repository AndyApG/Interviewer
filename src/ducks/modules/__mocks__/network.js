/* eslint-env jest */

import uuidv4 from '../../../utils/uuid';

const { actionTypes, entityPrimaryKeyProperty } = jest.requireActual('../network');

const reducer = jest.fn(() => ({
  nodes: [],
  edges: [],
  ego: {
    [entityPrimaryKeyProperty]: uuidv4(),
  },
}));

export default reducer;

export {
  actionTypes,
};
