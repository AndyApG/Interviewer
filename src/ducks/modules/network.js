import { reject, findIndex, isMatch, omit } from 'lodash';

import uuidv4 from '../../utils/uuid';

// Property name for the primary key for nodes
export const NodePrimaryKeyProperty = '_uid';

// Property name for node "model" properties
export const NodeAttributesProperty = 'attributes';

export const ADD_NODES = 'ADD_NODES';
export const REMOVE_NODE = 'REMOVE_NODE';
export const UPDATE_NODE = 'UPDATE_NODE';
export const TOGGLE_NODE_ATTRIBUTES = 'TOGGLE_NODE_ATTRIBUTES';
export const ADD_EDGE = 'ADD_EDGE';
export const TOGGLE_EDGE = 'TOGGLE_EDGE';
export const REMOVE_EDGE = 'REMOVE_EDGE';
export const SET_EGO = 'SET_EGO';
export const UNSET_EGO = 'UNSET_EGO';


// Initial network model structure
const initialState = {
  ego: {},
  nodes: [],
  edges: [],
};

function flipEdge(edge) {
  return { from: edge.to, to: edge.from, type: edge.type };
}

function edgeExists(edges, edge) {
  return (
    findIndex(edges, edge) !== -1 ||
    findIndex(edges, flipEdge(edge)) !== -1
  );
}

/**
 * existingNodes - Existing network.nodes
 * netNodes - nodes to be added to the network
 * additionalAttributes - node model properties
*/
function getNodesWithBatchAdd(existingNodes, newNodes, nodeAttributeData) {
  // Create a function to create a UUID and merge node attributes
  const withModelandAttributeData = newNode => ({
    [NodePrimaryKeyProperty]: uuidv4(),
    ...newNode, // second to allow existing UUID to be overwritten
    [NodeAttributesProperty]: {
      ...newNode[NodeAttributesProperty],
      ...nodeAttributeData,
    },
  });

  return existingNodes.concat(newNodes.map(withModelandAttributeData));
}

/**
 * @param {array} nodes - an array of objects representing nodes
 * @param {object} updatedNode - object representing the node to be updated and its new properties
 * @param {boolean} full - if the entire node should be overwritten
 */
function getUpdatedNodes(nodes, updatedNode, nodeAttributeData) {
  // Iterate over the nodes list
  const updatedNodes = nodes.map((node) => {
    // Skip nodes where the primary key doesn't match
    if (node[NodePrimaryKeyProperty] !== updatedNode[NodePrimaryKeyProperty]) { return node; }
    console.log('getUpdatedNodes');
    // if we have an attributes payload, merge with any existing attributes
    if (nodeAttributeData) {
      console.log('nodeAttributedata found');
      console.log(updatedNode, nodeAttributeData);
      return {
        ...node,
        ...updatedNode,
        [NodeAttributesProperty]: {
          ...node[NodeAttributesProperty],
          ...nodeAttributeData,
        },
        [NodePrimaryKeyProperty]: node[NodePrimaryKeyProperty],
      };
    }

    // If no attribute payload, just merge the new node data with the existing
    return {
      ...node,
      ...updatedNode,
      [NodePrimaryKeyProperty]: node[NodePrimaryKeyProperty],
    };
  });

  // Return the modified array of nodes
  return updatedNodes;
}

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case ADD_NODES: {
      return {
        ...state,
        nodes: getNodesWithBatchAdd(state.nodes, action.nodes, action.additionalAttributes),
      };
    }
    /**
     * TOGGLE_NODE_ATTRIBUTES
     * Used to toggle the value of a boolean variable between states when the node is tapped.
     */
    case TOGGLE_NODE_ATTRIBUTES: {
      // attributes = object containing node properties, minus _uid
      const attributes = omit(action.attributes, [NodePrimaryKeyProperty]);

      // Map over the nodes
      const updatedNodes = state.nodes.map((node) => {
        // Skip nodes with different primary keys
        if (node[NodePrimaryKeyProperty] !== action[NodePrimaryKeyProperty]) { return node; }

        //
        if (isMatch(node, attributes)) {
          return omit(node, Object.getOwnPropertyNames(attributes));
        }

        return {
          ...node,
          ...action.attributes,
        };
      });

      return {
        ...state,
        nodes: updatedNodes,
      };
    }
    case UPDATE_NODE: {
      return {
        ...state,
        nodes: getUpdatedNodes(state.nodes, action.node, action.additionalAttributes),
      };
    }
    case REMOVE_NODE: {
      const removeNodePrimaryKeyProperty = action[NodePrimaryKeyProperty];
      return {
        ...state,
        nodes: reject(state.nodes, node =>
          node[NodePrimaryKeyProperty] === removeNodePrimaryKeyProperty),
        edges: reject(state.edges, edge =>
          edge.from === removeNodePrimaryKeyProperty || edge.to === removeNodePrimaryKeyProperty),
      };
    }
    case ADD_EDGE:
      if (edgeExists(state.edges, action.edge)) { return state; }
      return {
        ...state,
        edges: [...state.edges, action.edge],
      };
    case TOGGLE_EDGE:
      // remove edge if it exists, add it if it doesn't
      if (edgeExists(state.edges, action.edge)) {
        return {
          ...state,
          edges: reject(reject(state.edges, action.edge), flipEdge(action.edge)),
        };
      }
      return {
        ...state,
        edges: [...state.edges, action.edge],
      };
    case REMOVE_EDGE:
      if (edgeExists(state.edges, action.edge)) {
        return {
          ...state,
          edges: reject(reject(state.edges, action.edge), flipEdge(action.edge)),
        };
      }
      return state;
    default:
      return state;
  }
}

/**
 * @namespace NetworkActionCreators
 */
const actionCreators = {};

const actionTypes = {
  ADD_NODES,
  UPDATE_NODE,
  TOGGLE_NODE_ATTRIBUTES,
  REMOVE_NODE,
  ADD_EDGE,
  TOGGLE_EDGE,
  REMOVE_EDGE,
  SET_EGO,
  UNSET_EGO,
};

export {
  actionCreators,
  actionTypes,
};
