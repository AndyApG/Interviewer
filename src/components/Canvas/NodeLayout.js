import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isEmpty, isEqual, pick, has } from 'lodash';
import LayoutNode from '../../containers/Canvas/LayoutNode';
import { nodePrimaryKeyProperty, getNodeAttributes, nodeAttributesProperty } from '../../ducks/modules/network';

const watchProps = ['width', 'height', 'rerenderCount'];

const propsChangedExcludingNodes = (nextProps, props) =>
  !isEqual(pick(nextProps, watchProps), pick(props, watchProps));

const nodesLengthChanged = (nextProps, props) =>
  nextProps.nodes.length !== props.nodes.length;

class NodeLayout extends Component {
  static propTypes = {
    nodes: PropTypes.array,
    onSelected: PropTypes.func.isRequired,
    connectFrom: PropTypes.string,
    highlightAttribute: PropTypes.string,
    allowPositioning: PropTypes.bool,
    canCreateEdge: PropTypes.bool,
    allowSelect: PropTypes.bool,
    layout: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
  };

  static defaultProps = {
    connectFrom: null,
    nodes: [],
    highlightAttribute: null,
    allowPositioning: true,
    allowSelect: true,
    canCreateEdge: true,
    layout: null,
    width: null,
    height: null,
  };

  shouldComponentUpdate(nextProps) {
    if (nodesLengthChanged(nextProps, this.props)) { return true; }
    if (propsChangedExcludingNodes(nextProps, this.props)) { return true; }

    return false;
  }

  isHighlighted(node) {
    return !isEmpty(this.props.highlightAttribute) &&
      node[nodeAttributesProperty][this.props.highlightAttribute] === true;
  }

  isLinking(node) {
    return this.props.canCreateEdge &&
      node[nodePrimaryKeyProperty] === this.props.connectFrom;
  }

  render() {
    const {
      nodes,
      allowPositioning,
      allowSelect,
      layout,
      width,
      height,
    } = this.props;

    return (
      <div className="node-layout">
        { nodes.map((node) => {
          const nodeAttributes = getNodeAttributes(node);
          if (!has(nodeAttributes, layout)) { return null; }

          return (
            <LayoutNode
              key={node[nodePrimaryKeyProperty]}
              node={node}
              layoutVariable={layout}
              onSelected={() => this.props.onSelected(node)}
              selected={this.isHighlighted(node)}
              linking={this.isLinking(node)}
              allowPositioning={allowPositioning}
              allowSelect={allowSelect}
              areaWidth={width}
              areaHeight={height}
            />
          );
        }) }
      </div>
    );
  }
}

export { NodeLayout };

export default NodeLayout;
