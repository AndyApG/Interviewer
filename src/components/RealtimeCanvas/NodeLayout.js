/* eslint-disable no-param-reassign */
import React from 'react';
import PropTypes from 'prop-types';
import { isEmpty, get } from 'lodash';
import LayoutContext from '../../contexts/LayoutContext';
import { entityPrimaryKeyProperty, entityAttributesProperty } from '../../ducks/modules/network';
import ScreenManager from './ScreenManager';
import LayoutNode from './LayoutNode';

class NodeLayout extends React.Component {
  constructor(props) {
    super(props);

    this.ref = React.createRef();
    this.updateRAF = undefined;
    this.layoutEls = [];
    this.screen = ScreenManager();
  }

  componentDidMount() {
    this.createLayoutEls();
    this.updateRAF = requestAnimationFrame(() => this.update());
  }

  componentDidUpdate(prevProps) {
    const { nodes } = this.props;

    if (prevProps.nodes.length !== nodes.length) {
      this.createLayoutEls();
    }
  }

  componentWillUnmount() {
    cancelAnimationFrame(this.updateRAF);
  }

  createLayoutEls = () => {
    const { nodes } = this.props;

    this.layoutEls = nodes.map((_, index) => {
      if (this.layoutEls[index]) { return this.layoutEls[index]; }

      const nodeEl = document.createElement('div');
      nodeEl.style.position = 'absolute';
      nodeEl.style.transform = 'translate(-50%, -50%)';
      nodeEl.style.display = 'none';
      this.ref.current.append(nodeEl);

      return nodeEl;
    });
  }

  update = () => {
    const {
      simulation: {
        getPosition,
      },
    } = this.context;

    this.layoutEls.forEach((el, index) => {
      const relativePosition = getPosition.current(index);
      if (!relativePosition || !el) { return; }

      const screenPosition = this.screen.calculateScreenCoords(relativePosition);
      el.style.left = `${screenPosition.x}px`;
      el.style.top = `${screenPosition.y}px`;
      el.style.display = 'block';
    });

    this.updateRAF = requestAnimationFrame(() => this.update());
  }

  isLinking = (node) => {
    const { connectFrom } = this.props;
    return get(node, entityPrimaryKeyProperty) === connectFrom;
  };

  isHighlighted = (node) => {
    const { highlightAttribute } = this.props;
    return (
      !isEmpty(highlightAttribute)
      && get(node, [entityAttributesProperty, highlightAttribute]) === true
    );
  };

  initializeLayout = (el) => {
    if (!el) { return; }
    this.ref.current = el;
    this.screen.initialize(el);
  };

  // (uuid, index, { dy, dx, x, y })
  handleDragStart = () => {};

  handleDragMove = (uuid, index, delta) => {
    const {
      network: { layout },
      simulation: { simulationEnabled, moveNode },
    } = this.context;

    const { updateNode } = this.props;

    const {
      dy,
      dx,
      x,
      y,
    } = delta;

    if (simulationEnabled) {
      moveNode({ dy, dx }, index);
      return;
    }

    updateNode(
      uuid,
      undefined,
      { [layout]: this.screen.calculateRelativeCoords({ x, y }) },
    );
  };

  handleDragEnd = (uuid, index, { x, y }) => {
    const {
      network: { layout },
      simulation: { simulationEnabled, releaseNode },
    } = this.context;

    const {
      updateNode,
    } = this.props;

    if (simulationEnabled) {
      releaseNode(index);
      return;
    }

    updateNode(
      uuid,
      undefined,
      { [layout]: this.screen.calculateRelativeCoords({ x, y }) },
    );
  };

  render() {
    const {
      network: { nodes },
      viewport,
      simulation: {
        reheat,
        stop,
        simulationEnabled,
        toggleSimulation,
      },
    } = this.context;

    const {
      allowPositioning,
      allowSelect,
      onSelected,
    } = this.props;

    return (
      <>
        <div className="node-layout" ref={this.initializeLayout} />

        <div style={{ position: 'absolute', top: 0, left: 0 }}>
          { !simulationEnabled && (
            <button type="button" onClick={() => toggleSimulation()}>enable simulation</button>
          )}
          { simulationEnabled && (
            <>
              <button type="button" onClick={() => toggleSimulation()}>disable simulation</button>
              <button type="button" onClick={() => reheat()}>reheat</button>
              <button type="button" onClick={() => stop()}>stop</button>
              <button type="button" onClick={() => viewport.zoomViewport(1.5)}>in</button>
              <button type="button" onClick={() => viewport.zoomViewport(0.67)}>out</button>
            </>
          )}
        </div>

        {nodes.map((node, index) => {
          const el = this.layoutEls[index];
          if (!el) { return null; }
          return (
            <LayoutNode
              node={node}
              portal={el}
              index={index}
              key={`${node[entityPrimaryKeyProperty]}_${index}`}
              onDragStart={this.handleDragStart}
              onDragMove={this.handleDragMove}
              onDragEnd={this.handleDragEnd}
              allowPositioning={allowPositioning}
              allowSelect={allowSelect}
              onSelected={onSelected}
              selected={this.isHighlighted(node)}
              linking={this.isLinking(node)}
            />
          );
        })}
      </>
    );
  }
}

NodeLayout.propTypes = {
  onSelected: PropTypes.func.isRequired,
  allowPositioning: PropTypes.bool,
  allowSelect: PropTypes.bool,
};

NodeLayout.defaultProps = {
  allowPositioning: true,
  allowSelect: true,
};

NodeLayout.contextType = LayoutContext;

export { NodeLayout };

export default NodeLayout;
