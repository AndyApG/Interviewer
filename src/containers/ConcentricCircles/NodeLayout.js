import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { compose, withHandlers, withState } from 'recompose';
import { isEqual, isEmpty, pick, isMatch, has } from 'lodash';
import LayoutNode from './LayoutNode';
import { withBounds } from '../../behaviours';
import { makeGetSociogramOptions, makeGetPlacedNodes } from '../../selectors/sociogram';
import { actionCreators as sessionsActions } from '../../ducks/modules/sessions';
import { nodePrimaryKeyProperty, getNodeAttributes } from '../../ducks/modules/network';
import { DropTarget } from '../../behaviours/DragAndDrop';
import sociogramOptionsProps from './propTypes';

const watchProps = ['width', 'height', 'dropCount'];

const propsChangedExcludingNodes = (nextProps, props) =>
  !isEqual(pick(nextProps, watchProps), pick(props, watchProps));

const nodesLengthChanged = (nextProps, props) =>
  nextProps.nodes.length !== props.nodes.length;

const relativeCoords = (container, node) => ({
  x: (node.x - container.x) / container.width,
  y: (node.y - container.y) / container.height,
});

const dropHandlers = compose(
  withState('dropCount', 'setDropCount', 0),
  withHandlers({
    accepts: () => ({ meta }) => meta.itemType === 'POSITIONED_NODE',
    onDrop: props => (item) => {
      props.updateNode(
        {
          [nodePrimaryKeyProperty]: item.meta[nodePrimaryKeyProperty],
        },
        {
          [props.layoutVariable]: relativeCoords(props, item),
        },
      );

      // Horrible hack for performance (only re-render nodes on drop, not on drag)
      props.setDropCount(props.dropCount + 1);
    },
    onDrag: props => (item) => {
      if (!has(item.meta, props.layoutVariable)) { return; }

      props.updateNode(
        {
          [nodePrimaryKeyProperty]: item.meta[nodePrimaryKeyProperty],
        },
        {
          [props.layoutVariable]: relativeCoords(props, item),
        },
      );
    },
  }),
);

class NodeLayout extends Component {
  static propTypes = {
    nodes: PropTypes.array,
    toggleEdge: PropTypes.func.isRequired,
    toggleHighlight: PropTypes.func.isRequired,
    ...sociogramOptionsProps,
  };

  static defaultProps = {
    nodes: [],
    allowPositioning: true,
    allowSelect: true,
  };

  constructor(props) {
    super(props);

    this.state = {
      connectFrom: null,
    };
  }

  shouldComponentUpdate(nextProps) {
    if (nodesLengthChanged(nextProps, this.props)) { return true; }
    if (propsChangedExcludingNodes(nextProps, this.props)) { return true; }

    return false;
  }

  onSelected = (node) => {
    const {
      allowSelect,
    } = this.props;

    if (!allowSelect) { return; }

    this.connectNode(node[nodePrimaryKeyProperty]);

    this.toggleHighlightAttributes(node[nodePrimaryKeyProperty]);

    this.forceUpdate();
  }

  connectNode(nodeId) {
    const { createEdge, canCreateEdge } = this.props;
    const { connectFrom } = this.state;

    if (!canCreateEdge) { return; }

    if (!connectFrom) {
      this.setState({ connectFrom: nodeId });
      return;
    }

    if (connectFrom !== nodeId) {
      this.props.toggleEdge({
        from: connectFrom,
        to: nodeId,
        type: createEdge,
      });
    }

    this.setState({ connectFrom: null });
  }

  toggleHighlightAttributes(nodeId) {
    if (!this.props.allowHighlighting) { return; }

    this.props.toggleHighlight(
      nodeId,
      { ...this.props.highlightAttributes },
    );
  }

  isHighlighted(node) {
    return !isEmpty(this.props.highlightAttributes) &&
      isMatch(node, this.props.highlightAttributes);
  }

  isLinking(node) {
    return this.props.allowSelect &&
      this.props.canCreateEdge &&
      node[nodePrimaryKeyProperty] === this.state.connectFrom;
  }

  render() {
    const {
      nodes,
      allowPositioning,
      allowSelect,
      layoutVariable,
    } = this.props;

    return (
      <div className="node-layout">
        { nodes.map((node) => {
          const nodeAttributes = getNodeAttributes(node);
          if (!has(nodeAttributes, layoutVariable)) { return null; }

          return (
            <LayoutNode
              key={node[nodePrimaryKeyProperty]}
              node={node}
              layoutVariable={layoutVariable}
              onSelected={() => this.onSelected(node)}
              selected={this.isHighlighted(node)}
              linking={this.isLinking(node)}
              allowPositioning={allowPositioning}
              allowSelect={allowSelect}
              areaWidth={this.props.width}
              areaHeight={this.props.height}
            />
          );
        }) }
      </div>
    );
  }
}

function makeMapStateToProps() {
  const getPlacedNodes = makeGetPlacedNodes();
  const getSociogramOptions = makeGetSociogramOptions();

  return function mapStateToProps(state, props) {
    return {
      nodes: getPlacedNodes(state, props),
      ...getSociogramOptions(state, props),
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    toggleHighlight: bindActionCreators(sessionsActions.toggleNodeAttributes, dispatch),
    toggleEdge: bindActionCreators(sessionsActions.toggleEdge, dispatch),
    updateNode: bindActionCreators(sessionsActions.updateNode, dispatch),
  };
}

export { NodeLayout };

export default compose(
  connect(makeMapStateToProps, mapDispatchToProps),
  withBounds,
  dropHandlers,
  DropTarget,
)(NodeLayout);
