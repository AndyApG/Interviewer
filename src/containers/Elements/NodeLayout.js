import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators, compose } from 'redux';
import { connect } from 'react-redux';
import { first, isMatch } from 'lodash';
import { Node } from 'network-canvas-ui';
import { draggable, withBounds, selectable } from '../../behaviours';
import { DropZone } from '../../components/Elements';
import { activePrompt } from '../../selectors/session';
import { actionCreators as networkActions } from '../../ducks/modules/network';

const label = node => node.nickname;

const draggableType = 'POSITIONED_NODE';

const EnhancedNode = draggable(selectable(Node));

export class NodeLayout extends Component {
  constructor(props) {
    super(props);

    this.state = {
      connectFrom: null,
    };
  }

  onSelectNode = (node) => {
    if (!this.props.prompt.selectAction) { return; }

    switch (this.props.prompt.selectAction) {
      case 'EDGE':
        this.connectNode(node); break;
      case 'ATTRIBUTES':
        this.toggleNodeAttributes(node); break;
      default:
    }
  };

  updateNodeLayout = (hits, coords, node) => {
    const { prompt, updateNode } = this.props;
    const hit = first(hits);
    const relativeCoords = {
      x: (coords.x - hit.x) / hit.width,
      y: (coords.y - hit.y) / hit.height,
    };

    updateNode({ ...node, [prompt.layout]: relativeCoords });
  }

  connectNode(node) {
    const nodeId = node.id;
    const edgeType = this.props.prompt.edgeType;
    const connectFrom = this.state.connectFrom;


    if (!connectFrom) {
      this.setState({ connectFrom: nodeId });
      return;
    }

    if (connectFrom !== nodeId) {
      this.props.toggleEdge({
        from: connectFrom,
        to: nodeId,
        type: edgeType,
      });
    }

    this.setState({ connectFrom: null });
  }

  toggleNodeAttributes(node) {
    this.props.toggleNodeAttributes(node, this.props.prompt.nodeAttributes);
  }

  isSelected(node) {
    switch (this.props.prompt.selectAction) {
      case 'EDGE':
        return (node.id === this.state.connectFrom);
      case 'ATTRIBUTES':
        return isMatch(node, this.props.prompt.nodeAttributes);
      default:
        return false;
    }
  }

  render() {
    const { prompt, nodes, width, height } = this.props;

    const canDrag = Object.prototype.hasOwnProperty.call(prompt, 'canDrag') ? prompt.canDrag : true;
    const canSelect = Object.prototype.hasOwnProperty.call(prompt, 'canSelect') ? prompt.canSelect : true;

    return (
      <DropZone droppableName="NODE_LAYOUT" acceptsDraggableType={draggableType}>
        <div className="node-layout">
          { nodes.map((node, key) => {
            if (!Object.prototype.hasOwnProperty.call(node, prompt.layout)) { return null; }

            const x = node[prompt.layout].x * width;
            const y = node[prompt.layout].y * height;

            return (
              <div key={key} className="node-layout__node" style={{ left: `${x}px`, top: `${y}px` }}>
                <EnhancedNode
                  label={label(node)}
                  draggableType={draggableType}
                  onDropped={(hits, coords) => this.updateNodeLayout(hits, coords, node)}
                  onDrag={(hits, coords) => this.updateNodeLayout(hits, coords, node)}
                  onSelected={() => this.onSelectNode(node)}
                  selected={this.isSelected(node)}
                  canDrag={canDrag}
                  canSelect={canSelect}
                  animate={false}
                  {...node}
                />
              </div>
            );
          }) }
        </div>
      </DropZone>
    );
  }
}

NodeLayout.propTypes = {
  nodes: PropTypes.array,
  updateNode: PropTypes.func.isRequired,
  toggleEdge: PropTypes.func.isRequired,
  toggleNodeAttributes: PropTypes.func.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  prompt: PropTypes.object.isRequired,
};

NodeLayout.defaultProps = {
  nodes: [],
};

function mapStateToProps(state) {
  return {
    prompt: activePrompt(state),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    updateNode: bindActionCreators(networkActions.updateNode, dispatch),
    toggleNodeAttributes: bindActionCreators(networkActions.toggleNodeAttributes, dispatch),
    toggleEdge: bindActionCreators(networkActions.toggleEdge, dispatch),
  };
}

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withBounds,
)(NodeLayout);
