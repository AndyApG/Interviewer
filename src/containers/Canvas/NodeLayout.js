import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { compose, withHandlers, withState } from 'recompose';
import { has } from 'lodash';
import { withBounds } from '../../behaviours';
import { actionCreators as sessionsActions } from '../../ducks/modules/sessions';
import { nodePrimaryKeyProperty, nodeAttributesProperty } from '../../ducks/modules/network';
import { DropTarget } from '../../behaviours/DragAndDrop';
import NodeLayout from '../../components/Canvas/NodeLayout';
import { makeGetPlacedNodes } from '../../selectors/canvas';

const relativeCoords = (container, node) => ({
  x: (node.x - container.x) / container.width,
  y: (node.y - container.y) / container.height,
});

const withRerenderCount = withState('rerenderCount', 'setRerenderCount', 0);

const withDropHandlers = withHandlers({
  accepts: () => ({ meta }) => meta.itemType === 'POSITIONED_NODE',
  onDrop: ({ updateNode, layout, setRerenderCount, rerenderCount, width, height, x, y }) =>
    (item) => {
      updateNode(
        item.meta,
        {
          [layout]: relativeCoords({ width, height, x, y }, item),
        },
      );

      // Horrible hack for performance (only re-render nodes on drop, not on drag)
      setRerenderCount(rerenderCount + 1);
    },
  onDrag: ({ layout, updateNode, width, height, x, y }) => (item) => {
    if (!has(item.meta[nodeAttributesProperty], layout)) { return; }

    updateNode(
      item.meta,
      {
        [layout]: relativeCoords({ width, height, x, y }, item),
      },
    );
  },
  onDragEnd: ({ layout, setRerenderCount, rerenderCount }) => (item) => {
    if (!has(item.meta[nodeAttributesProperty], layout)) { return; }

    // make sure to also re-render nodes that were updated on drag end
    setRerenderCount(rerenderCount + 1);
  },
});

const withSelectHandlers = compose(
  withHandlers({
    connectNode: ({ createEdge, connectFrom, updateLinkFrom, toggleEdge }) =>
      (nodeId) => {
        if (!createEdge) { return; }

        if (!connectFrom) {
          updateLinkFrom(nodeId);
          return;
        }

        if (connectFrom !== nodeId) {
          toggleEdge({
            from: connectFrom,
            to: nodeId,
            type: createEdge,
          });
        }

        updateLinkFrom(null);
      },
    toggleHighlightAttribute: ({ allowHighlighting, highlight, toggleHighlight }) =>
      (node) => {
        if (!allowHighlighting) { return; }
        const newVal = !node[nodeAttributesProperty][highlight];
        toggleHighlight(
          node[nodePrimaryKeyProperty],
          { [highlight]: newVal },
        );
      },
  }),
  withHandlers({
    onSelected: ({
      allowSelect,
      connectNode,
      toggleHighlightAttribute,
      setRerenderCount,
      rerenderCount,
    }) => (node) => {
      if (!allowSelect) { return; }

      connectNode(node[nodePrimaryKeyProperty]);
      toggleHighlightAttribute(node);
      setRerenderCount(rerenderCount + 1);
    },
  }),
);

function makeMapStateToProps() {
  const getPlacedNodes = makeGetPlacedNodes();

  return function mapStateToProps(state, { createEdge, allowHighlighting, subject, layout }) {
    const allowSelect = createEdge || allowHighlighting;

    return {
      nodes: getPlacedNodes(state, { subject, layout }),
      allowSelect,
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    toggleHighlight: bindActionCreators(sessionsActions.toggleNodeAttributes, dispatch),
    toggleEdge: bindActionCreators(sessionsActions.toggleEdge, dispatch),
    updateNode: bindActionCreators(sessionsActions.updateNode, dispatch),
  };
}

export default compose(
  connect(makeMapStateToProps, mapDispatchToProps),
  withBounds,
  withRerenderCount,
  withDropHandlers,
  withSelectHandlers,
  DropTarget,
)(NodeLayout);
