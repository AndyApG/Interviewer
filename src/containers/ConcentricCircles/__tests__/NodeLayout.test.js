/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { NodeLayout } from '../NodeLayout';
import { NodePrimaryKeyProperty } from '../../../ducks/modules/network';

const layout = 'foo';

const mockProps = {
  nodes: [
    { bar: 'buzz', [NodePrimaryKeyProperty]: 123, [layout]: { x: 0, y: 0 } },
  ],
  updateNode: () => {},
  toggleEdge: () => {},
  toggleHighlight: () => {},
  width: 123,
  height: 456,
  layoutVariable: layout,
  createEdge: 'bar',
  displayEdges: [],
  canCreateEdge: false,
  canHighlight: false,
  highlightAttributes: {},
  allowHighlighting: false,
  allowPositioning: false,
  selectMode: '',
  sortOrderBy: [],
  concentricCircles: 0,
  skewedTowardCenter: false,
};

describe('<NodeLayout />', () => {
  it('renders ok', () => {
    const component = shallow(<NodeLayout {...mockProps} />);

    expect(component).toMatchSnapshot();
  });

  it('shouldComponentUpdate() when prop `nodes` changes length only', () => {
    const componentDidUpdate = jest.fn();
    NodeLayout.prototype.componentDidUpdate = componentDidUpdate;

    const component = shallow(
      <NodeLayout {...mockProps} />,
      { lifecycleExperimental: true },
    );

    component.setProps({
      ...mockProps,
      nodes: [{ bar: 'buzz', [NodePrimaryKeyProperty]: 123 }],
    });

    component.setProps({
      ...mockProps,
      nodes: [{ bar: 'buzz', [NodePrimaryKeyProperty]: 123 }, { bar: 'bing', [NodePrimaryKeyProperty]: 456 }],
    });

    expect(componentDidUpdate.mock.calls.length).toEqual(1);
  });

  it('shouldComponentUpdate() when prop other than `nodes` changes', () => {
    const componentDidUpdate = jest.fn();
    NodeLayout.prototype.componentDidUpdate = componentDidUpdate;

    const component = shallow(
      <NodeLayout {...mockProps} />,
      { lifecycleExperimental: true },
    );

    component.setProps({
      ...mockProps,
      nodes: [{ bar: 'buzz', [NodePrimaryKeyProperty]: 123 }],
    });

    component.setProps({
      ...mockProps,
      width: 789,
    });

    expect(componentDidUpdate.mock.calls.length).toEqual(1);
  });
});
