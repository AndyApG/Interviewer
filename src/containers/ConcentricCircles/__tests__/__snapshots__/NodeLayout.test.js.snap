// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeLayout /> renders ok 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NodeLayout
    allowHighlighting={false}
    allowPositioning={false}
    allowSelect={true}
    canCreateEdge={false}
    canHighlight={false}
    concentricCircles={0}
    createEdge="bar"
    displayEdges={Array []}
    height={456}
    highlightAttributes={Object {}}
    layoutVariable="foo"
    nodes={
      Array [
        Object {
          "_uid": 123,
          "bar": "buzz",
          "foo": Object {
            "x": 0,
            "y": 0,
          },
        },
      ]
    }
    selectMode=""
    skewedTowardCenter={false}
    sortOrder={Array []}
    toggleEdge={[Function]}
    toggleHighlight={[Function]}
    updateNode={[Function]}
    width={123}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <LayoutNode
          allowPositioning={false}
          allowSelect={true}
          areaHeight={456}
          areaWidth={123}
          layoutVariable="foo"
          linking={false}
          node={
            Object {
              "_uid": 123,
              "bar": "buzz",
              "foo": Object {
                "x": 0,
                "y": 0,
              },
            }
          }
          onSelected={[Function]}
          selected={false}
        />,
      ],
      "className": "node-layout",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "123",
        "nodeType": "class",
        "props": Object {
          "allowPositioning": false,
          "allowSelect": true,
          "areaHeight": 456,
          "areaWidth": 123,
          "layoutVariable": "foo",
          "linking": false,
          "node": Object {
            "_uid": 123,
            "bar": "buzz",
            "foo": Object {
              "x": 0,
              "y": 0,
            },
          },
          "onSelected": [Function],
          "selected": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <LayoutNode
            allowPositioning={false}
            allowSelect={true}
            areaHeight={456}
            areaWidth={123}
            layoutVariable="foo"
            linking={false}
            node={
              Object {
                "_uid": 123,
                "bar": "buzz",
                "foo": Object {
                  "x": 0,
                  "y": 0,
                },
              }
            }
            onSelected={[Function]}
            selected={false}
          />,
        ],
        "className": "node-layout",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "123",
          "nodeType": "class",
          "props": Object {
            "allowPositioning": false,
            "allowSelect": true,
            "areaHeight": 456,
            "areaWidth": 123,
            "layoutVariable": "foo",
            "linking": false,
            "node": Object {
              "_uid": 123,
              "bar": "buzz",
              "foo": Object {
                "x": 0,
                "y": 0,
              },
            },
            "onSelected": [Function],
            "selected": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
